### ISSUE TITLE
StreamPeerTCP disconnects after one received response

### ISSUE THREAD

**Issue by Gamemap** [godotengine/godot#62001]
### Godot version

v4.0.alpha10.official [4bbe7f0b9]

### System information

Windows 11 64 Bit

### Issue description

I'm trying to connect to an xmpp server (Prosody) to create a chat bot, but I can't authenticate because after receiving the first response from the server the StreamPeerTCP node disconnects.
The following script produces this output (the lines with only one number are the print commands with `get_available_bytes()`):

```
--- Debugging process started ---
Godot Engine v4.0.alpha8.official.cc3ed63af - https://godotengine.org
Vulkan API 1.2.0 - Using Vulkan Device #0: NVIDIA - NVIDIA GeForce GTX 1660 Ti
 
Status: 2
198
STATUS: 2
Response: <?xml version='1.0'?><stream:stream id='2bdde09e-00d9-4756-9dd9-0807c76d8761' xmlns:stream='http://etherx.jabber.org/streams' from='bot.example.test' xmlns='jabber:component:accept'></stream:stream>  STATUS: 2
Status: 2
-1
Status: 0
Status: 0
Status: 0
Status: 0
Status: 0
Status: 0
Status: 0
--- Debugging process stopped ---
```

Also this Error message appers two times (for each get_available_bytes() function  in the script one time)
```
E 0:00:00:0617   _process: Condition "!is_open()" is true. Returning: -1
  <C++-Quellcode>drivers/unix/net_socket_posix.cpp:730 @ get_available_bytes()
  <Stacktrace>   xmpp.gd:27 @ _process()
```

### Steps to reproduce

I have connected to a Windows Prosody server using the script below (configuration below the script). Because the Windows versions have security issues, Prosody does not offer a Windows download. Here is an alternative download link: https://de.freedownloadmanager.org/Windows-PC/Prosody-KOSTENLOS.html
The bug appears also on the prosody linux version.

**Script:**
```
extends Control


var tcp_peer = StreamPeerTCP.new()

var server_ip = "127.0.0.1"

var port = 5347


var stanza = '<stream:stream xmlns="jabber:component:accept" xmlns:stream="http://etherx.jabber.org/streams" id="af15442e-24bc-4239-a8cd-274a5569adce" from="bot.example.test" to="bot.example.test" xml:lang="en"/>'


func _ready():
	tcp_peer.connect_to_host(server_ip,port)
	tcp_peer.poll()
	
	tcp_peer.put_data(str(stanza).to_utf8_buffer())
	pass

func _process(_delta):
	print("Status: ",tcp_peer.get_status())
	if (tcp_peer.get_status() == StreamPeerTCP.STATUS_CONNECTED):
		if tcp_peer.has_method("poll"):
			tcp_peer.poll()
		
		print(tcp_peer.get_available_bytes())
		
		
		if  tcp_peer.get_available_bytes()>0:
			print("STATUS: ",tcp_peer.get_status())
			var response = tcp_peer.get_string(tcp_peer.get_available_bytes())
			print("Response: ",response,"  STATUS: ",tcp_peer.get_status())

```

**Prosody Config:**
```
-- Prosody Example Configuration File
--
-- Information on configuring Prosody can be found on our
-- website at http://prosody.im/doc/configure
--
-- Tip: You can check that the syntax of this file is correct
-- when you have finished by running: luac -p prosody.cfg.lua
-- If there are any errors, it will let you know what and where
-- they are, otherwise it will keep quiet.
--
-- The only thing left to do is rename this file to remove the .dist ending, and fill in the
-- blanks. Good luck, and happy Jabbering!


---------- Server-wide settings ----------
-- Settings in this section apply to the whole server and are the default settings
-- for any virtual hosts

-- This is a (by default, empty) list of accounts that are admins
-- for the server. Note that you must create the accounts separately
-- (see http://prosody.im/doc/creating_accounts for info)
-- Example: admins = { "user1@example.com", "user2@example.net" }
admins = { }

-- Enable use of libevent for better performance under high load
-- For more information see: http://prosody.im/doc/libevent
--use_libevent = true;

-- This is the list of modules Prosody will load on startup.
-- It looks for mod_modulename.lua in the plugins folder, so make sure that exists too.
-- Documentation on modules can be found at: http://prosody.im/doc/modules
modules_enabled = {

	-- Generally required
		"roster"; -- Allow users to have a roster. Recommended ;)
		"saslauth"; -- Authentication for clients and servers. Recommended if you want to log in.
		--"tls"; -- Add support for secure TLS on c2s/s2s connections
		"dialback"; -- s2s dialback support
		"disco"; -- Service discovery

	-- Not essential, but recommended
		"private"; -- Private XML storage (for room bookmarks, etc.)
		"vcard"; -- Allow users to set vCards
	
	-- These are commented by default as they have a performance impact
		--"privacy"; -- Support privacy lists
		--"compression"; -- Stream compression

	-- Nice to have
		"version"; -- Replies to server version requests
		"uptime"; -- Report how long server has been running
		"time"; -- Let others know the time here on this server
		"ping"; -- Replies to XMPP pings with pongs
		"pep"; -- Enables users to publish their mood, activity, playing music and more
		"register"; -- Allow users to register on this server using a client and change passwords

	-- Admin interfaces
		"admin_adhoc"; -- Allows administration via an XMPP client that supports ad-hoc commands
		--"admin_telnet"; -- Opens telnet console interface on localhost port 5582
	
	-- HTTP modules
		--"bosh"; -- Enable BOSH clients, aka "Jabber over HTTP"
		--"http_files"; -- Serve static files from a directory over HTTP

	-- Other specific functionality
		--"groups"; -- Shared roster support
		--"announce"; -- Send announcement to all online users
		--"welcome"; -- Welcome users who register accounts
		--"watchregistrations"; -- Alert admins of registrations
		--"motd"; -- Send a message to users when they log in
		--"legacyauth"; -- Legacy authentication. Only used by some old clients and bots.
};

-- These modules are auto-loaded, but should you want
-- to disable them then uncomment them here:
modules_disabled = {
	-- "offline"; -- Store offline messages
	-- "c2s"; -- Handle client connections
	-- "s2s"; -- Handle server-to-server connections
};

-- Disable account creation by default, for security
-- For more information see http://prosody.im/doc/creating_accounts
allow_registration = true;

-- These are the SSL/TLS-related settings. If you don't want
-- to use SSL/TLS, you may comment or remove this
--ssl = {
--	key = "certs/localhost.key";
--	certificate = "certs/localhost.crt";
--}

-- Force clients to use encrypted connections? This option will
-- prevent clients from authenticating unless they are using encryption.

c2s_require_encryption = false

-- Force certificate authentication for server-to-server connections?
-- This provides ideal security, but requires servers you communicate
-- with to support encryption AND present valid, trusted certificates.
-- NOTE: Your version of LuaSec must support certificate verification!
-- For more information see http://prosody.im/doc/s2s#security

s2s_secure_auth = false

-- Many servers don't support encryption or have invalid or self-signed
-- certificates. You can list domains here that will not be required to
-- authenticate using certificates. They will be authenticated using DNS.

--s2s_insecure_domains = { "gmail.com" }

-- Even if you leave s2s_secure_auth disabled, you can still require valid
-- certificates for some domains by specifying a list here.

--s2s_secure_domains = { "jabber.org" }

-- Select the authentication backend to use. The 'internal' providers
-- use Prosody's configured data storage to store the authentication data.
-- To allow Prosody to offer secure authentication mechanisms to clients, the
-- default provider stores passwords in plaintext. If you do not trust your
-- server please see http://prosody.im/doc/modules/mod_auth_internal_hashed
-- for information about using the hashed backend.

authentication = "internal_plain"

-- Select the storage backend to use. By default Prosody uses flat files
-- in its configured data directory, but it also supports more backends
-- through modules. An "sql" backend is included by default, but requires
-- additional dependencies. See http://prosody.im/doc/storage for more info.

--storage = "sql" -- Default is "internal"

-- For the "sql" backend, you can uncomment *one* of the below to configure:
--sql = { driver = "SQLite3", database = "prosody.sqlite" } -- Default. 'database' is the filename.
--sql = { driver = "MySQL", database = "prosody", username = "prosody", password = "secret", host = "localhost" }
--sql = { driver = "PostgreSQL", database = "prosody", username = "prosody", password = "secret", host = "localhost" }

-- Logging configuration
-- For advanced logging see http://prosody.im/doc/logging
log = {
	-- info = "prosody.log"; -- Change 'info' to 'debug' for verbose logging
	-- error = "prosody.err";
	info = "*console"; -- Log to the console, useful for debugging with daemonize=false
}

----------- Virtual hosts -----------
-- You need to add a VirtualHost entry for each domain you wish Prosody to serve.
-- Settings under each VirtualHost entry apply *only* to that host.

VirtualHost "localhost"

VirtualHost "example.test"


	-- Assign this host a certificate for TLS, otherwise it would use the one
	-- set in the global section (if any).
	-- Note that old-style SSL on port 5223 only supports one certificate, and will always
	-- use the global one.
	--ssl = {
	--	key = "certs/example.test.key";
	--	certificate = "certs/example.test.crt";
	--}

------ Components ------
-- You can specify components to add hosts that provide special services,
-- like multi-user conferences, and transports.
-- For more information on components, see http://prosody.im/doc/components

---Set up a MUC (multi-user chat) room server on conference.example.com:
Component "conference.example.test" "muc"

-- Set up a SOCKS5 bytestream proxy for server-proxied file transfers:
--Component "proxy.example.com" "proxy65"

---Set up an external component (default component port is 5347)
--
-- External components allow adding various services, such as gateways/
-- transports to other networks like ICQ, MSN and Yahoo. For more info
-- see: http://prosody.im/doc/components#adding_an_external_component
--
Component "bot.example.test"
	component_secret = "1234"

```

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/46293.

**Comment by Gamemap** [comment]
I know you have a lot to do for Godot 4.0, but would it be possible that someone could look at this so I can continue to work on this program?

**Comment by Gamemap** [comment]
@Faless Could you try this and tell me if this is a problem on my side or a bug? 
(I don't mean to be rude, so if you don't have time to look at this, I can understand it.)

**Comment by Faless** [comment]
@Gamemap StreamPeerTCP is widely used in Godot codebase, including in the HTTPClient which handles keep-alive (multiple requests without disconnecting).

One issue I can see in your code is in the `_ready` where you `poll` once and expect to be connected immediately putting data, which might not (and likely will not) be the case in a real internet scenario.

You should keep polling while the status is [STATUS_CONNECTING](https://docs.godotengine.org/en/latest/classes/class_streampeertcp.html#enum-streampeertcp-status), and only after reaching `STATUS_CONNECTED` put the data.

You should also check the return value of `put_data` to see if you got an error.

**Comment by Gamemap** [comment]
@Faless Thank you for your quick reply. I will change my code ann in a few days (when I'm back at my computer).

**Comment by Gamemap** [comment]
@Faless I think I found my error. The Prosody server disconnects the client after sending a response with an ID, so I have to reconnect to the server and send a new request with the given ID (I couldn't verify this, but I think this might be the problem).

I will close this report and comment again if I can verify my theory.

Thank you very much for helping me and informing me about how to correctly use `poll`.
